Re-authentication

bigmontz opened the pull request:
**⚠️ This API is released as preview.**

This changes introduce two ways of changing the connection credentials in a driver instance, each of them solving a different use case. 

### Token Expiration / Change Credentials for the whole driver instance

This use case is related to the issue #993 in the repository. For solving this, the driver is now able to receive a `AuthTokenManager` in the driver creation. This interface enables the user code provide new auth tokens to the driver and be notified by token expiration failures. 

For simplifying the usage, the driver also provides a default implementation of `AuthTokenManager` which can be created with `neo4j. expirationBasedAuthTokenManager` and receives a function for renewing the auth token as parameters. 

Example:

```typescript
import neo4j, { AuthToken } from 'neo4j-driver'

/**
 * Method called whenever the driver needs to refresh the token.
 *
 * The refresh will happen if the driver is notified by the server
 * about a token expiration or if the `Date.now() > tokenData.expiry`
 *
 * Important, the driver will block all the connections creation until
 * this function resolves the new auth token.
 */
async function fetchAuthTokenFromMyProvider () {
   const bearer: string = await myProvider.getBearerToken()
   const token: AuthToken = neo4j.auth.bearer(bearer)
   const expiration: Date = myProvider.getExpiryDate()  
   return {
      token,
      // if expiration is not provided, 
      // the driver will only fetch a new token when a failure happens
      expiration 
   }
}

const driver = neo4j.driver(
    'neo4j://localhost:7687', 
    neo4j.expirationBasedAuthTokenManager({ 
        tokenProvider: fetchAuthTokenFromMyProvider 
    })
)
```

### User Switching

In this scenario, different credentials can be configured in a session providing a way for change the user context for the session. For using this feature, it needed to check if your server supports session auth by calling `driver.supportsSessionAuth()`.

Example:

```typescript
import neo4j from 'neo4j-driver'


const driver =  neo4j.driver(
    'neo4j://localhost:7687', 
    neo4j.auth.basic('neo4j', 'password')
)


const sessionWithUserB = driver.session({
  database: 'neo4j',
  auth: neo4j.auth.basic('userB', 'userBpassword')
})


try {
  // run some queries as userB
  const result = await sessionWithUserB.executeRead(tx => tx.run('RETURN 1'))
} finally {
  // close the session as usual
  await sessionWithUserB.close()
}
```

**⚠️ This API is released as preview.**

bigmontz committed:
Add AuthTokenProvider configuration

bigmontz committed:
Handling auth error

bigmontz committed:
Add support for impersonation

bigmontz committed:
Add holder

bigmontz committed:
Re-Authentication

bigmontz committed:
Fix tests

bigmontz committed:
Adjust test

bigmontz committed:
Add login/logoff concepts

bigmontz committed:
Add basic testkit tests for AuthProvider

bigmontz committed:
Ajust the re-auth flow and support FakeTime in tk

bigmontz committed:
Session config

bigmontz committed:
Pool: separate validation on acquire and on release and add the acquisition context to create and validate

bigmontz committed:
Close connection, pipeline re-auth and passing to test_renewable_auth.TestRenewableAuth5x1 tests

bigmontz committed:
Ajust user switching

bigmontz committed:
Add tests to DenoJS

bigmontz committed:
Testing pool changes

bigmontz committed:
Add 5.1 base protocol test and supportLogoff test

bigmontz committed:
Add test for login/logoff in a non supported protocol

bigmontz committed:
Add test to the connection-channel

bigmontz committed:
Add tests for AuthenticationProvider#authenticate, no user-switching

bigmontz committed:
Add AuthenticationProvider.authenticate when auth is provider

bigmontz committed:
Test AuthenticationProvider.handleError

bigmontz committed:
Add tests for direct provider error handle

bigmontz committed:
Add tests for handle error in the RoutingConnectionProvider

bigmontz committed:
Add test for 5.1 in RoutingProvider

bigmontz committed:
Add test for RoutingConnectionProvider resource creation setup (pool)

bigmontz committed:
Add tests to the RoutingConnectionProvider pool configuration

bigmontz committed:
Add tests to the DirectConnectionProvider pool configuration

bigmontz committed:
Small fix and sync deno

bigmontz committed:
Fix merge issues

bigmontz committed:
Implement user-switch doesn't allow sticky connection

bigmontz committed:
Improve tests

bigmontz committed:
Improve tests around non-sticky connections

bigmontz committed:
Mark connections created with a new auth and without support to re-auth as sticky

bigmontz committed:
Killing sticky connections on-release

bigmontz committed:
Enable require a new connection from the pool

bigmontz committed:
Acquire new sticky connection if backwards compatible and protocol doesn't support re-auth

bigmontz committed:
Add `supportsSessionAuth`

bigmontz committed:
Add CheckSessionAuthSupport to testkit

bigmontz committed:
Add tests for session configuration

bigmontz committed:
Add user switching to RxSession

bigmontz committed:
VerifyAuthorization

bigmontz committed:
Add backwardsCompatibleAuth flag to the driver configuration

bigmontz committed:
VerifyAuth: refresh auth if needed

bigmontz committed:
Add AuthTokenManager

bigmontz committed:
Adjust testkit

bigmontz committed:
Release sticky connection and  only drop connection when needed

bigmontz committed:
Fix pool size comparison

bigmontz committed:
VerifyAuthentication ajusts

The auth validation when the connection is not new should be done in the provider.

bigmontz committed:
Clean and test

bigmontz committed:
sync deno

bigmontz committed:
Implement changes from testkit

bigmontz committed:
Add temporal token tests

ConorNeo commented on packages/bolt-connection/test/bolt/bolt-protocol-v5x1.test.js:
```diff
@@ -0,0 +1,1092 @@
+/**
```
Not for this PR but i wonder if we could have these work off each other somehow, on the server we have Annotations that say run this test on x,y,z versions to stop duplication

bigmontz replied:
I don't bother to have duplication in the test code, it make it easier to address differences and make the tests totally decoupled.

Maybe it worth to add subset of re-usable tests per capability, it can be a interesting re-factory to do.

ConorNeo commented on packages/bolt-connection/test/bolt/bolt-protocol-v4x0.test.js:
```diff
@@ -589,6 +589,15 @@ describe('#unit BoltProtocolV4x0', () => {
     })
   })
+
+  describe('Bolt v5.1', () => {
```
Do we not need the other message (login)

bigmontz replied:
I only need to check outside the protocol level if driver supports logoff (which means also it support logon). I might be the case I just called `supportReauth`.

ConorNeo commented on packages/bolt-connection/src/bolt/bolt-protocol-v5x1.js:
```diff
@@ -0,0 +1,132 @@
+   *
+   * @returns {StreamObserver} the stream observer that monitors the corresponding server response.
+   */
+  login ({ authToken, onComplete, onError, flush } = {}) {
```
The Server Calls this Logon, i don't know what the other drivers do and if it should be logon or login but i suspect either the server should change or the drivers so we all align nicely

bigmontz replied:
I've started with login because the driver already uses observer called `LoginObserver`, but it is fair enough keep the message with same name as in the server since it will make easier to people read the logs and follow the code, for example.

bigmontz committed:
rename login => logon

bigmontz committed:
rename supportsLogoff => supportsReAuth

bigmontz committed:
Enable AuthTokenManager tests for testkit

bigmontz committed:
Marking re-auth and user-switch as experimental/preview

bigmontz committed:
Retry on token expired when token is static token

bigmontz committed:
Adjust documentation

bigmontz committed:
Remove backwardsCompatibilityAuth code

bigmontz committed:
Fix rebase issues

bigmontz committed:
Rename TemporalAuthTokenManager to ExpirationBasedAuthTokenManager

Also rename the structure with token and expiration.

bigmontz committed:
Fix names in testkit-backend

bigmontz committed:
Ajust docs

bigmontz committed:
testkit-backend: Fix `authToken` param in VerifyAuthentication

bigmontz committed:
Remove file added by acident

bigmontz committed:
testkit-backend: Adjust authToken name in
VerifyAuthentication

bigmontz committed:
Sync DenoJS

bigmontz committed:
PolyfillFlatMap

bigmontz committed:
polyfill flatmap

tristanls commented:
Hi, apologies if this is the wrong place to ask. Thank you, been looking forward to this work. It looks like this missed the 5.7 release. Is it feasible to put out 5.8 release so we can test out this API preview?

ConorNeo commented:
Hi, yes this is merged into dev and (unless something changes) will be in 5.8

jhanggi commented:
I know this feature is in preview, so happy to wait. I wanted to try this out since we've had problems in the past with `Neo4jError: LDAP authorization info expired`. We've worked around it by catching that particular error and retrying, but this felt like it might be a cleaner solution.

I believe the typescript types may not all be updated. It seems the second argument to `neo4j.driver` still only accepts AuthToken.

`Argument of type 'AuthTokenManager' is not assignable to parameter of type 'AuthToken'. Type 'AuthTokenManager' is missing the following properties from type 'AuthToken': scheme, credentials`

```
const driver = neo4j.driver(
  'neo4j://localhost:7687',
  neo4j.expirationBasedAuthTokenManager({
    // tokenProvider rather than getAuthData
    tokenProvider: () => {
      return neo4j.basic('username', 'password')
    }),
  }),
);
```

bigmontz commented:
@jhanggi, thanks for spotting the issue. 

I've updated the type definitions in the PR: https://github.com/neo4j/neo4j-javascript-driver/pull/1089

I've also updated description of this PR with the correct example.

bigmontz commented:
Fix released in 5.9.0
